File: cli/promptforge_cli/archive.py
"""Extract a downloaded kit archive into a target directory."""

from __future__ import annotations

import shutil
import tempfile
import zipfile
from pathlib import Path


def extract_zip_stripped(zip_path: Path, target: Path) -> None:
    """Extract ``zip_path`` into ``target``, stripping a single top-level folder.

    GitHub archives wrap everything in one ``<repo>-<ref>/`` directory; we drop
    that so the code lands directly in ``target``.
    """
    target.mkdir(parents=True, exist_ok=True)
    with tempfile.TemporaryDirectory() as tmp:
        tmp_path = Path(tmp)
        with zipfile.ZipFile(zip_path) as zf:
            zf.extractall(tmp_path)
        entries = list(tmp_path.iterdir())
        src = entries[0] if len(entries) == 1 and entries[0].is_dir() else tmp_path
        for item in src.iterdir():
            dest = target / item.name
            if dest.exists():
                if dest.is_dir():
                    shutil.rmtree(dest)
                else:
                    dest.unlink()
            shutil.move(str(item), str(dest))

File: vscode-extension/src/archive.ts
import * as fs from "fs";
import * as path from "path";

import AdmZip from "adm-zip";

/**
 * Extract a zip buffer into ``target``, stripping the single top-level folder
 * that GitHub archives wrap everything in.
 */
export function extractZipStripped(zipBuffer: Buffer, target: string): void {
  const zip = new AdmZip(zipBuffer);
  const entries = zip.getEntries();
  if (entries.length === 0) return;

  // GitHub archives share one "<repo>-<ref>/" prefix; detect and drop it.
  const first = entries[0].entryName;
  const slashIndex = first.indexOf("/");
  const prefix = slashIndex > 0 ? first.slice(0, slashIndex + 1) : "";

  for (const entry of entries) {
    if (entry.isDirectory) continue;
    const relative =
      prefix && entry.entryName.startsWith(prefix)
        ? entry.entryName.slice(prefix.length)
        : entry.entryName;
    if (!relative) continue;
    const dest = path.join(target, relative);
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.writeFileSync(dest, entry.getData());
  }
}
